### PR TITLE
Flow Affinity via Environment & Agent Labels

### PR DESCRIPTION
**Pull request by joshmeek**
**Thanks for contributing to Prefect!**

Please describe your work and make sure your PR:

- [x] adds new tests (if appropriate)
- [x] updates `CHANGELOG.md` (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

Note that your PR will not be reviewed unless all three boxes are checked.

## What does this PR change?
Adds label specification to execution environments as well as agents.

## Why is this PR important?
When you want to run multiple agents and specify which flows get deployed by which agent there needs to be a mechanism in place for this.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,12 +6,11 @@ These changes are available in the [master branch](https://github.com/PrefectHQ/
 
 ### Features
 
-- None
+- Environments and Agents now support labeling for execution specification - [#1651](https://github.com/PrefectHQ/prefect/pull/1651)
```
Comment (jlowin) [inline CHANGELOG.md]:
```suggestion
- Flow Affinity: Environments and Agents now support labeling for execution specification - [#1651](https://github.com/PrefectHQ/prefect/pull/1651)
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -22,7 +22,6 @@ These changes are available in the [master branch](https://github.com/PrefectHQ/
 ### Fixes
 
 - Fix defaults for unspecified ARNs in the Fargate Agent - [#1634](https://github.com/PrefectHQ/prefect/pull/1634)
-- Fix ShellTask return value on empty stdout - [#1632](https://github.com/PrefectHQ/prefect/pull/1632)
```
Comment (cicdw) [inline CHANGELOG.md]: why did this get removed?

File: src/prefect/agent/agent.py [hunk 3]
```diff
@@ -38,10 +38,15 @@ class Agent:
...
         - name (str, optional): An optional name to give this agent. Can also be set through
             the environment variable `PREFECT__CLOUD__AGENT__NAME`. Defaults to "agent"
+        - labels (List[str], optional): a list of labels, which are arbitrary string identifiers used by Prefect
+            Agents when polling for work
     """
 
-    def __init__(self, name: str = None) -> None:
+    def __init__(self, name: str = None, labels: Iterable[str] = None) -> None:
         self.name = name or config.cloud.agent.get("name", "agent")
+        self.labels = labels or config.cloud.agent.get("labels")
```
Comment (cicdw) [inline src/prefect/agent/agent.py]: have you tested this? I suspect you'll need to add:
```python
import ast

ast.literal_eval(config.cloud.agent.get("labels", 'None')) # note the quotes around None
```
but I'm not sure how smart our configuration object is when inferring lists from environment variables.

File: tests/agent/test_agent.py [hunk 4]
```diff
@@ -17,6 +17,7 @@ def test_agent_init(runner_token):
 def test_agent_config_options(runner_token):
     with set_temporary_config({"cloud.agent.auth_token": "TEST_TOKEN"}):
         agent = Agent()
+        assert agent.labels is None
```
Comment (cicdw) [inline tests/agent/test_agent.py]: I'm surprised this test passes when the default appears to be `[]`

File: CHANGELOG.md (hunk 2 above)
Comment (joshmeek) [reply CHANGELOG.md]: I have no idea, my changelog is really messed up for some reason

**Pull request merged**